src/consensus/aft/raft.h

```diff
@@ -1355,29 +1354,53 @@ namespace aft
         }
       }
 
-      send_append_entries_response(from, AppendEntriesResponseType::OK);
+      send_append_entries_response_ack(from, r);
     }
 
-    void send_append_entries_response(
-      ccf::NodeId to,
-      AppendEntriesResponseType answer,
-      const std::optional<ccf::TxID>& rejected = std::nullopt)
+    void send_append_entries_response_ack(
+      ccf::NodeId to, const AppendEntries& ae)
     {
-      aft::Index response_idx = state->last_idx;
-      aft::Term response_term = state->current_view;
+      // If we get to here, we have applied up to r.idx in this AppendEntries.
+      // We must only ACK this far, as we know nothing about the agreement of a
+      // suffix we may still hold _after_ r.idx with the leader's log
+      const auto response_idx = ae.idx;
+      send_append_entries_response(
+        to, AppendEntriesResponseType::OK, state->current_view, response_idx);
+    }
 
-      if (answer == AppendEntriesResponseType::FAIL && rejected.has_value())
+    void send_append_entries_response_nack(
```

Comment:
imho clearer as two:
send_append_entries_response_nack(ccf::NodeId to)
send_append_entries_response_nack(ccf::NodeId to, const ccf::TxID& rejected)